Transfer model

ifullgaz opened the pull request:
Transfer model

Depends on #2

ifullgaz committed:
common/model: add transfer model

ifullgaz committed:
common/model: fix typo in transfer finalize

ifullgaz committed:
WIP: transfer transactions iterator

ifullgaz committed:
entangled/transfer: added a reset function for transaction and completed transfer iterator

ifullgaz committed:
entangled/transfer: fix to bundle hash calculation

ifullgaz committed:
entangled/model: Fix erroneous memory initialization value

ifullgaz committed:
entangled/model: Fix erroneous memory initialization fields

ifullgaz committed:
entangled/model: Fix typos, added transaction reset test, remove init of address fields

ifullgaz committed:
entangled/model: rewrite of *free functions

ifullgaz committed:
entangled/model: fix bundle hash calculation - added bundle hash calculation test

ifullgaz committed:
entangled/model: fixed memory leak in test, optimized transaction serialize/deserialize

ifullgaz committed:
entangled/transfer: added signature function pointer support

ifullgaz committed:
entangled/transaction: rewrite to support flex trits

th0br0 commented:
@ifullgaz could you rebase this one as well please

jserv commented on common/model/transfer.h:
```diff
@@ -0,0 +1,243 @@
+/*
+ * Copyright (c) 2018 IOTA Stiftung
+ * https://gitlab.com/iota-foundation/software/entangled
```
Ditto.

jserv reviewed (commented):
Fix inconsistent Gitlab link

ifullgaz committed:
entangled/trinary: added conversion function from/to trytes to/from flex_trits
entangled/model: added cpp wrapper class around iota_transaction structure
entangled/cppclient: rewrite to use cpptransaction and flexible trits

ifullgaz committed:
entangled/model: added missing test file

ifullgaz committed:
entangled/model: added namespace iota, capital letter for CPP Transaction class

ifullgaz committed:
entangled/trinary: cpp wrapper for flex_trit

ifullgaz committed:
entangled/trinary: cpp wrapper for flex_trit

ifullgaz committed:
entangled: applied clang_format

ifullgaz committed:
entangled/trinary: bug fixes, added trit access to trytes, cleanup BUILD files
entangled/model: added serialize to cpptransaction, added tests for cpptransaction, cleanup BUILD files

ifullgaz committed:
entangled/trinary: Added 3:1 tests, fixed transaction serialize for 3:1 storage scheme

ifullgaz committed:
entangled/model: Disabled 3:1 tests

ifullgaz committed:
entangled/model: Renamed cpptransaction to TryteTransaction, cleanup BUILD files,

th0br0 commented on common/model/tests/BUILD:
```diff
@@ -6,3 +6,13 @@ cc_test(
+cc_test (
+    name = "test_transfer",
+    srcs = ["test_transfer.c"],
+    linkopts = ["-lpthread"],
```
Why `-lpthread`?

th0br0 commented on common/model/tests/test_transfer.c:
```diff
@@ -0,0 +1,39 @@
+void test_bundle_hash(coid) {
+  transfer_t transfer = transfer_new(VALUE_IN);
+  transfer_value_in_t value_in = transfer_value_in(transfer);
+  transfer_value_in_set_address(value_in, (const tryte_t *)"9XV9RJGFJJZWITDPKSQXRTHCKJAIZZY9BYLBEQUXUNCLITRQDR9CCD99AANMXYEKD9GLJGVB9HIAGRIBQ");
```
please run `clang-format` - did you install the hooks?

th0br0 commented on common/model/tests/test_transfer.c:
```diff
@@ -0,0 +1,39 @@
+  transfer_t transfer = transfer_new(VALUE_IN);
+  transfer_value_in_t value_in = transfer_value_in(transfer);
+  transfer_value_in_set_address(value_in, (const tryte_t *)"9XV9RJGFJJZWITDPKSQXRTHCKJAIZZY9BYLBEQUXUNCLITRQDR9CCD99AANMXYEKD9GLJGVB9HIAGRIBQ");
+  transfer_set_tag(transfer, (tryte_t *)"ZTDIDNQDJZGUQKOWJ9JZRCKOVGP");
```
either cast to const here as well or don't at all?

th0br0 commented on common/model/transaction.c:
```diff
@@ -23,96 +23,89 @@
+  transaction_set_address(transaction, position);
+  position += sizeof(transaction->address);
+  transaction_set_value(transaction, trytes_to_long(position, 27));
+  position += 27;
```
could we get rid of the magic numbers here?

th0br0 commented on common/model/transaction.c:
```diff
@@ -23,96 +23,89 @@
-  memcpy(trytes + offset, transaction_obsolete_tag(transaction),
+  position += sizeof(transaction->address);
+  long_to_trytes(transaction_value(transaction), position);
+  position += 27;
```
magic numbers ...

th0br0 reviewed (changes requested):
Rebase and fix against master please

ifullgaz committed:
entangled/model: Transfer model using flex_trit.
entangled/helpers: added flex_trit signature functions.
entangled/kerl: renamed functions trits_to_bytes and bytes_to_trits to convert_trits_to_bytes and convert_bytes_to_trits

ifullgaz committed:
entangled/helpers: fix test not compiling

th0br0 commented on common/helpers/BUILD:
```diff
@@ -19,6 +19,49 @@ cc_library(
+)
+
+cc_library(
+    name = "sign_3",
```
you don't need all these targets, just depend on `trit_array` directly?

ifullgaz replied:
It allows testing the signing functions with each memory model.

th0br0 commented on common/model/transfer.c:
```diff
@@ -0,0 +1,715 @@
+// Set the seed (copy argument)
+void transfer_output_set_seed(transfer_output_t transfer_output,
+                              const flex_trit_t *seed) {
+  memcpy(transfer_output->seed, seed, sizeof(transfer_output->seed));
```
We shouldn't copy here

ifullgaz replied:
transfer_output_t defines seed as an array of 243 flex trits. seed cannot be directly assigned to transfer_output->seed, it has can only be copied into the array.
With a new approach to keep pointers to data, it is possible to assign instead of copy.

th0br0 commented on common/model/transfer.c:
```diff
@@ -0,0 +1,715 @@
+// struct _transfer_output {...};
+
+// Creates and returns a new transfer output
+transfer_output_t transfer_output_new(void) {
```
I think you confused the output/input naming

th0br0 replied:
this is for a `value_in`(put)

ifullgaz replied:
Following a discussion we had a few months ago, we agreed that the naming for in and out should be "swapped".

th0br0 commented on common/model/transfer.c:
```diff
@@ -0,0 +1,715 @@
+}
+
+// Get the index
+size_t transfer_output_index(transfer_output_t transfer_output) {
```
let's call this `seed_index`

th0br0 commented on common/model/transfer.c:
```diff
@@ -0,0 +1,715 @@
+}
+
+// Get the tag
+flex_trit_t *transfer_tag(transfer_t transfer) { return transfer->tag; }
```
the tag should be per part of the transfer, not global?

ifullgaz replied:
I prefer not to use globals.

th0br0 commented on common/model/transfer.c:
```diff
@@ -0,0 +1,715 @@
+int transfer_ctx_hash(transfer_ctx_t transfer_ctx, Kerl *kerl,
+                      transfer_t *transfers, size_t len) {
+  size_t i, j, count, current_index = 0;
+  tryte_t essence[TRYTES_PER_ESSENCE];
```
why do you need tryte here?

ifullgaz replied:
In the 4:1 and 5:1 models, we would need to use the 'insert' function which would be far less efficient. When kerl support flex_trit natively, it will become advantageous since we can remove the conversion from trytes (or flex_trits) to trits also.

th0br0 reviewed (commented):
Transfers should use as little memory as possible most of the time. 

Could you change the approach so that the only thing that the transfer objects keep around is pointers to the actual addresses/seeds/...?

th0br0 reviewed (changes requested):
Transfers should use as little memory as possible most of the time. 

Could you change the approach so that the only thing that the transfer objects keep around is pointers to the actual addresses/seeds/...?

ifullgaz committed:
entangled/model: Transfer model redesigned to hold pointers to data instead of copy of data

th0br0 commented on common/model/transfer.c:
```diff
@@ -0,0 +1,673 @@
+#define TRITS_PER_MESSAGE 6561
+#define TRITS_PER_ESSENCE 486
+#define TRITS_PER_BUNDLE_HASH 243
+#define TRYTES_PER_MESSAGE 2187
```
It appears you're using these as offset multiplicators; however, you're using flextrits! That won't work.

th0br0 commented on common/model/tryte_transaction.cc:
```diff
@@ -50,7 +50,7 @@ std::string TryteTransaction::message(void) {
+// Set the transaction message (assign)
 void TryteTransaction::setMessage(const std::string &trytes) {
   size_t flex_len = flex_trits_num_for_trits(NUM_TRITS_SIGNATURE);
   flex_trit_t trits[flex_len];
```
All that the `transaction_set_message` methods do is memcpy.
Can't you just memset & copy straight into the transaction's structure? 
All of these methods waste memory

ifullgaz committed:
common/helpers: cleanup BUILD files, added IOTA_EXPORT for flex functions
common/model: added transfer BUILD, fix transfer iterator, optimised tryte_transaction, added transfer iterator test
common/storage: added TEST_TRANSACTION for each memory model

ifullgaz committed:
common/storage: fix sqlite tests for 4:1 and 5:1 memory schemes

thibault-martinez commented on common/helpers/sign.c:
```diff
@@ -72,5 +73,66 @@ char* iota_sign_signature_gen(char const* const seed, size_t const index,
 
+IOTA_EXPORT flex_trit_t* iota_flex_sign_address_gen(
+    flex_trit_t const* const seed, size_t const index, size_t const security) {
+  if (!(security > 0 && security <= 3)) return NULL;
```
add brackets

thibault-martinez commented on common/helpers/sign.c:
```diff
@@ -72,5 +73,66 @@ char* iota_sign_signature_gen(char const* const seed, size_t const index,
+  Kerl kerl;
+
+  trit_t subseed[HASH_LENGTH];
+  trit_t key[key_length];
```
use calloc for variable length static arrays (for windows)

thibault-martinez commented on common/helpers/sign.c:
```diff
@@ -72,5 +73,66 @@ char* iota_sign_signature_gen(char const* const seed, size_t const index,
+  kerl_reset(&kerl);
+
+  int8_to_flex_trit_array(address, HASH_LENGTH, key, key_length, HASH_LENGTH);
+  memset(key, 0, key_length * sizeof(trit_t));
```
Is it really useful ?

ifullgaz replied:
I remember reading somewhere that it's good practice, to avoid "guessing" the key from the "noise" generated by the function. It was there before, I've left it untouched.

thibault-martinez replied:
Interesting, thanks

thibault-martinez commented on common/helpers/sign.c:
```diff
@@ -72,5 +73,66 @@ char* iota_sign_signature_gen(char const* const seed, size_t const index,
+  trit_t subseed[HASH_LENGTH];
+  trit_t key[key_length];
+  flex_trit_t* address =
+      calloc(flex_trits_num_for_trits(HASH_LENGTH) + 1, sizeof(flex_trit_t));
```
return NULL if out of memory

thibault-martinez commented on common/helpers/sign.c:
```diff
@@ -72,5 +73,66 @@ char* iota_sign_signature_gen(char const* const seed, size_t const index,
+
+IOTA_EXPORT flex_trit_t* iota_flex_sign_signature_gen(
+    flex_trit_t const* const seed, size_t const index, size_t const security,
+    flex_trit_t const* const bundleHash) {
```
bundle_hash

thibault-martinez commented on common/helpers/sign.c:
```diff
@@ -72,5 +73,66 @@ char* iota_sign_signature_gen(char const* const seed, size_t const index,
+
+  trit_t hash[HASH_LENGTH];
+  trit_t subseed[HASH_LENGTH];
+  trit_t key[key_length];
```
calloc

thibault-martinez commented on common/helpers/sign.c:
```diff
@@ -72,5 +73,66 @@ char* iota_sign_signature_gen(char const* const seed, size_t const index,
+  trit_t subseed[HASH_LENGTH];
+  trit_t key[key_length];
+  flex_trit_t* signature =
+      calloc(flex_trits_num_for_trits(key_length) + 1, sizeof(flex_trit_t));
```
return NULL if out of memory

thibault-martinez commented on common/helpers/sign.c:
```diff
@@ -72,5 +73,66 @@ char* iota_sign_signature_gen(char const* const seed, size_t const index,
+
+  iss_kerl_signature(key, hash, key, key_length, &kerl);
+  int8_to_flex_trit_array(signature, key_length, key, key_length, key_length);
+  memset(key, 0, key_length * sizeof(trit_t));
```
useful ?

ifullgaz replied:
I remember reading somewhere that it's good practice, to avoid "guessing" the key from the "noise" generated by the function. It was there before, I've left it untouched.

thibault-martinez commented on common/model/transaction.c:
```diff
@@ -414,6 +414,22 @@ void transaction_set_hash(iota_transaction_t transaction,
+         sizeof(transaction->branch));
+  memset(transaction->tag, FLEX_TRIT_NULL_VALUE, sizeof(transaction->tag));
+  memset(transaction->nonce, FLEX_TRIT_NULL_VALUE, sizeof(transaction->nonce));
+}
```
What about the address and hash ?

ifullgaz replied:
Following a discussion with Andreas, it was decided to leave them untouched.

ifullgaz committed:
common/helpers: use calloc instead of variable length arrays. return NULL if out of memory. Rename bundleHash to bundle_hash.

ifullgaz committed:
common/model: remove lpthread linkopt.

ifullgaz committed:
Rebase, migrate to separate flex_trit
